Refs #23968 -- Removed unnecessary lists, generators, and tuple calls.

jdufresne opened the pull request:
<https://code.djangoproject.com/ticket/23968>

charettes commented:
I wonder if it could be worth adding [flake8-comprehensions](https://pypi.python.org/pypi/flake8-comprehensions/1.2.1) to CI to prevent new ones from slipping in.

jdufresne commented:
Thanks for highlighting this package! I wasn't aware of it.

It does show a bit of additional cleanup that would should do. I've pasted below. If we decide to go through with adding the package, should all cleanups be done in this commit?

One issue, [migrations do not serialize sets as set literals](https://github.com/django/django/blob/f04a40491764bdc9a2ebbfc03fa7be424fb3ce63/django/db/migrations/serializer.py#L245-L248) to more easily handle the edge case of an empty set. To handle this I guess we would either need to:

* Change migrations to use set literals and handle the edge case
* Change generated migrations manually when commiting
* Add some ignore rules for migrations

What do you think?

```
$ flake8
./django/contrib/contenttypes/migrations/0001_initial.py:32:29: C405 Unnecessary list literal - rewrite as a set literal.
./django/contrib/gis/gdal/raster/source.py:253:12: C407 Unnecessary list comprehension - 'sum' can take a generator.
./django/contrib/auth/hashers.py:381:18: C402 Unnecessary generator - rewrite as a dict comprehension.
./django/contrib/auth/migrations/0001_initial.py:29:36: C405 Unnecessary list literal - rewrite as a set literal.
./django/contrib/redirects/migrations/0001_initial.py:33:36: C405 Unnecessary list literal - rewrite as a set literal.
./django/views/i18n.py:242:30: C402 Unnecessary generator - rewrite as a dict comprehension.
./django/utils/cache.py:68:14: C402 Unnecessary generator - rewrite as a dict comprehension.
./django/utils/cache.py:97:10: C402 Unnecessary generator - rewrite as a dict comprehension.
./django/forms/boundfield.py:50:16: C400 Unnecessary generator - rewrite as a list comprehension.
./django/db/migrations/autodetector.py:1068:27: C402 Unnecessary generator - rewrite as a dict comprehension.
./django/db/migrations/autodetector.py:1072:27: C402 Unnecessary generator - rewrite as a dict comprehension.
./django/db/models/base.py:1297:18: C400 Unnecessary generator - rewrite as a list comprehension.
./django/db/models/sql/subqueries.py:55:30: C407 Unnecessary list comprehension - 'tuple' can take a generator.
./django/db/models/sql/query.py:803:29: C407 Unnecessary list comprehension - 'tuple' can take a generator.
./django/db/models/sql/query.py:804:23: C407 Unnecessary list comprehension - 'tuple' can take a generator.
./scripts/manage_translations.py:117:17: C407 Unnecessary list comprehension - 'sorted' can take a generator.
./scripts/manage_translations.py:145:28: C407 Unnecessary list comprehension - 'sorted' can take a generator.
./docs/_ext/djangodocs.py:250:37: C407 Unnecessary list comprehension - 'sum' can take a generator.
./tests/view_tests/tests/test_debug.py:316:17: C400 Unnecessary generator - rewrite as a list comprehension.
./tests/migrations/test_migrations_initial_false/0001_not_initial.py:28:29: C405 Unnecessary list literal - rewrite as a set literal.
./tests/migrations/test_migrations_fake_split_initial/0001_initial.py:28:29: C405 Unnecessary list literal - rewrite as a set literal.
./tests/migrations/test_migrations/0001_initial.py:33:29: C405 Unnecessary list literal - rewrite as a set literal.
./tests/model_meta/tests.py:165:30: C407 Unnecessary list comprehension - 'sorted' can take a generator.
./tests/model_meta/tests.py:246:13: C407 Unnecessary list comprehension - 'sorted' can take a generator.
./tests/model_meta/tests.py:256:13: C407 Unnecessary list comprehension - 'sorted' can take a generator.
./tests/schema/tests.py:172:25: C407 Unnecessary list comprehension - 'all' can take a generator.
./tests/gis_tests/gdal_tests/test_srs.py:183:25: C407 Unnecessary list comprehension - 'all' can take a generator.
./tests/gis_tests/gdal_tests/test_srs.py:184:25: C407 Unnecessary list comprehension - 'all' can take a generator.
./tests/gis_tests/distapp/tests.py:419:40: C407 Unnecessary list comprehension - 'sorted' can take a generator.
./tests/delete/tests.py:189:45: C400 Unnecessary generator - rewrite as a list comprehension.
./tests/m2m_through_regress/test_multitable.py:43:30: C407 Unnecessary list comprehension - 'sorted' can take a generator.
./tests/model_inheritance/test_abstract_inheritance.py:322:20: C400 Unnecessary generator - rewrite as a list comprehension.
./tests/prefetch_related/tests.py:861:30: C407 Unnecessary list comprehension - 'sorted' can take a generator.
```

adamchainz commented:
As author of `flake8-comprehensions`, I like the idea 👌

charettes commented:
@jdufresne I think changing `SetSerializer` to use literals when possible makes the most sense here.

jdufresne commented:
> I think changing SetSerializer to use literals when possible makes the most sense here.

I tend to agree. I've proposed this in PR #8436.

jdufresne commented:
I have updated the commit to include all fixes recommended by flake8-comprehensions except those surrounding generated migrations files.

I'm now also referencing old ticket [#23968](https://code.djangoproject.com/ticket/23968), as it looks to be the same goal as this change. But I can make a newer cleanup ticket if other prefer.

The tool was extremely useful in finding code to convert. However, [a false positive](https://github.com/adamchainz/flake8-comprehensions/issues/48) makes it difficult to integrate the tool with Django at this time. I think if that bug gets fixed, we could reconsider it.

ngnpope commented:
@jdufresne: could we not add the following lines near to the top of generated migration modules to avoid issues with generated code triggering style errors?

```python
# flake8: noqa
# isort: skip_file
```

(I think these are the options - but have tried to remember them off the top of my head...)

This way flake8-comprehensions could be integrated in CI.

The disadvantage is that it disables all checks on these files so not even things like syntax errors will be highlighted. Unfortunately flake8 still does not allow specifying a selection of checks in this comment.

adamchainz commented:
Disabling all checks on a file to add a new lint plugin is unacceptable, especially when the lint plugin was authored by a Django core member who will merge the fix to the plugin to prevent the false positive tomorrow 😉

adamchainz commented:
`flake8-comprehensions` 1.3.0 is out now with @jdufresne 's fix for false positives, I'd include it now

timgraham commented on django/db/models/sql/query.py:
```diff
@@ -800,8 +800,8 @@ def change_aliases(self, change_map):
         if isinstance(self.group_by, tuple):
-            self.group_by = tuple([col.relabeled_clone(change_map) for col in self.group_by])
-        self.select = tuple([col.relabeled_clone(change_map) for col in self.select])
+            self.group_by = tuple(col.relabeled_clone(change_map) for col in self.group_by)
```
Apparently using a list comprehension performs better here, see https://github.com/django/django/pull/7730/files#r93793194. If someone can explain when exactly that's the case, we can add it to the coding style guidelines in https://github.com/django/django/pull/8284.

adamchainz replied:
Ah yes, I remember that. It seems to be generally the case, even with a range of 100k items:

```
In [1]: %timeit tuple(x for x in range(100000))
100 loops, best of 3: 9.75 ms per loop

In [2]: %timeit tuple([x for x in range(100000)])
100 loops, best of 3: 7.04 ms per loop
```

The difference *is* tiny, the main point of #7730 and my related PR's is to optimize the often repeated code paths like `_clone`, not these individual operations. I changed it there just because of the comment.

Anyway we could just turn off the rule in `flake8-comprehensions`

jdufresne replied:
Oh wow. That is surprising and counter to my intuition. Thanks for pointing that out. Judging by the `timeit` output, I suppose the guideline ought to be that `tuple(<list comprehension>)` is normally preferred over `tuple(<generator expression>)`. I find them both very readable, but one is ever slightly faster. What do you think?

jdufresne commented:
I have updated the commit.

* Rebased to latest mater
* Reverted `tuple(<LIST COMPREHENSION>)` pattern until a more concrete decision is made on which pattern is preferred
* Using newest version of flake8-comprehension, caught additional uses of `tuple()` and `dict()` that can be literals instead

adamchainz commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -250,7 +250,7 @@ def geotransform(self):
     def geotransform(self, values):
         "Set the geotransform for the data source."
-        if sum([isinstance(x, (int, float)) for x in values]) != 6:
+        if sum(isinstance(x, (int, float)) for x in values) != 6:
```
this is just plain wrong, it allows through `[1,2,3,4,5,6,'foo']`. It should really be `len(values) == 6 and all(isinstance(x, (int, float) for x in values)`

adamchainz replied:
Will fix this in another PR

adamchainz replied:
#8569

adamchainz commented on tests/backends/tests.py:
```diff
@@ -970,9 +970,7 @@ def runner():
-        self.assertEqual(
-            len(set(conn.connection for conn in connections_dict.values())),
-            3)
+        self.assertEqual(len({conn.connection for conn in connections_dict.values()}), 3)
```
could leave this as multiple lines

jdufresne replied:
This line is 89 characters long. I was under the impression the Django code doesn't push hard for an 80 character length limit, but instead an 100-120 character limit. This isn't even the longest line in this same file.

From https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/

> An exception to PEP 8 is our rules on line lengths. Don’t limit lines of code to 79 characters if it means the code looks significantly uglier or is harder to read. We allow up to 119 characters as this is the width of GitHub code review; anything longer requires horizontal scrolling which makes review more difficult. This check is included when you run flake8. Documentation, comments, and docstrings should be wrapped at 79 characters, even though PEP 8 suggests 72.

adamchainz replied:
Sure, I just found the original version more readable. There is a soft limit on 79 chars meaning if it looks fine you can break.

adamchainz commented on tests/generic_views/test_base.py:
```diff
@@ -136,9 +136,8 @@ def test_invalid_keyword_argument(self):
-            kwargs = dict(((method, "value"),))
             with self.assertRaises(TypeError):
-                SimpleView.as_view(**kwargs)
+                SimpleView.as_view(method='value')
```
this isn't the same! you should do `**{method: 'value'}`

jdufresne replied:
Sorry, maybe I'm missing something. How are these not equivalent other than style? Both forms are supplying a single keyword argument named `method` with the value `'value'`, right?

adamchainz replied:
No, the dict's key is not `'method'` but whatever is in the variable `method` 😉

jdufresne replied:
Ah! Of course. i totally misread that. Will fix.

jdufresne commented:
Thanks for the review. Integrated feedback from @adamchainz

* Deferred changing questionable code in `django/contrib/gis/gdal/raster/source.py `
* Fixed my previous mixup of key names in ` tests/generic_views/test_base.py `
